File: src/main/java/com/baidu/algorithm/datastructure/_146_LRU_Cache.py
class LRUCache(object):
    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.mp = {}
        self.li = []

    def put_key_to_tail(self, key):
        self.li.remove(key)
        self.li.append(key)

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key in self.mp:
            self.put_key_to_tail(key)
            return self.mp[key]

        return -1

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: void
        """
        if key in self.mp:
            self.put_key_to_tail(key)
            self.mp[key] = value
        else:
            self.li.append(key)
            if len(self.li) > self.capacity:
                del_key = self.li.pop(0)
                del self.mp[del_key]
        self.mp[key] = value

File: src/main/java/com/baidu/algorithm/datastructure/_146_LRU_Cache.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.datastructure;

import java.util.HashMap;
import java.util.Map;

/**
 * _146_LRU_Cache
 *
 * @author xuhaoran01
 */
public class _146_LRU_Cache {

    class LRUCache {

        class Node {
            public int key;
            public int val;
            public Node next;
            public Node prev;

            public Node(int x, int y) {
                key = x;
                val = y;
            }
        }

        private Map<Integer, Node> key2Node;
        private int capacity;
        private int curSize;
        private Node head, tail;

        public LRUCache(int capacity) {
            key2Node = new HashMap<>();
            this.capacity = capacity;
            curSize = 0;

            head = new Node(-1, -1);
            tail = new Node(-1, -1);
            head.next = tail;
            tail.prev = head;
        }

        public int get(int key) {
            if(key2Node.containsKey(key)) {
                Node node = key2Node.get(key);
                deleteNode(node);

                putToHead(node);

                return node.val;
            }

            return -1;
        }

        public void put(int key, int value) {
            Node node = null;
            if(key2Node.containsKey(key)) {
                node = key2Node.get(key);
                deleteNode(node);

                node.val = value;
                putToHead(node);
            }
            else {
                if(curSize >= capacity) {
                    key2Node.remove(tail.prev.key);
                    deleteNode(tail.prev);
                }

                node = new Node(key, value);
                key2Node.put(key, node);

                putToHead(node);
                curSize++;
            }
        }

        private void deleteNode(Node node) {
            node.next.prev = node.prev;
            node.prev.next = node.next;
        }

        private void putToHead(Node node) {
            node.next = head.next;
            head.next.prev = node;

            head.next = node;
            node.prev = head;
        }
    }

    public static void main(String[] args) {
        LRUCache obj = new _146_LRU_Cache().new LRUCache(1);
        obj.put(2, 1);
        obj.get(2);
        obj.put(3, 2);
        obj.get(2);
        obj.get(3);
    }
}
